File: TiaCodegen/Commands/Functions/CTUCall.cs
using System.Linq;
using TiaCodegen.Commands.Functions.Base;
using TiaCodegen.Enums;
using TiaCodegen.Interfaces;

namespace TiaCodegen.Commands.Functions
{
    public class CTUCall : SystemFunctionBlockCall
    {
        public CTUCall(string instanceName,
            IOperationOrSignal cu = null,
            IOperationOrSignal r = null,
            IOperationOrSignal pv = null,
            IOperationOrSignal q = null,
            IOperationOrSignal cv = null) : base("CTU", instanceName, null)
        {
            Interface["CU"] = new IOperationOrSignalDirectionWrapper(cu, Direction.Input);
            Interface["R"] = new IOperationOrSignalDirectionWrapper(r, Direction.Input);
            Interface["PV"] = new IOperationOrSignalDirectionWrapper(pv, Direction.Input);
            Interface["Q"] = new IOperationOrSignalDirectionWrapper(q, Direction.Output);
            Interface["CV"] = new IOperationOrSignalDirectionWrapper(cv, Direction.Output);

            TemplateValueName = "value_type";
            TemplateValueType = "Type";
            TemplateValue = "Int";
            HasNoEn = true;

            Children.AddRange(Interface.Values.Where(x => x.OperationOrSignal != null).Select(x => x.OperationOrSignal));
        }
    }
}

File: TiaCodegen-ts/src/Commands/Functions/CTUCall.ts
import { IOperationOrSignal } from '../../Interfaces/IOperationOrSignal.js';
import { IOperationOrSignalDirectionWrapper } from '../../Interfaces/IOperationOrSignalDirectionWrapper.js';
import { Direction } from '../../Enums/Direction.js';
import { SystemFunctionBlockCall } from './Base/SystemFunctionBlockCall.js';

export interface CTUCallOptions {
    instanceName: string;
    r?: IOperationOrSignal | null;
    pv?: IOperationOrSignal | null;
    q?: IOperationOrSignal | null;
    cv?: IOperationOrSignal | null;
    templateValue?: string;
}

export class CTUCall extends SystemFunctionBlockCall {
    constructor(options: CTUCallOptions) {
        const { instanceName, r = null, pv = null, q = null, cv = null, templateValue = 'Int' } = options;
        super({ functionName: 'CTU', instanceName });
        this.iface['R'] = new IOperationOrSignalDirectionWrapper(r, Direction.Input);
        this.iface['PV'] = new IOperationOrSignalDirectionWrapper(pv, Direction.Input);
        this.iface['Q'] = new IOperationOrSignalDirectionWrapper(q, Direction.Output);
        this.iface['CV'] = new IOperationOrSignalDirectionWrapper(cv, Direction.Output);

        this.templateValueName = 'value_type';
        this.templateValueType = 'Type';
        this.templateValue = templateValue;
        this.hasNoEn = true;

        for (const w of Object.values(this.iface)) {
            if (w.operationOrSignal !== null) this.children.push(w.operationOrSignal);
        }
    }
}
